photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinate.scala

```diff
@@ -95,60 +90,76 @@ protected[ml] class FixedEffectCoordinate[Objective <: DistributedObjectiveFunct
   override protected[algorithm] def score(model: DatumScoringModel): CoordinateDataScores = model match {
 
     case fixedEffectModel: FixedEffectModel =>
-      FixedEffectCoordinate.score(dataset, fixedEffectModel)
+      FixedEffectCoordinate.score(dataset, fixedEffectModel, featureShardId)
 
     case _ =>
       throw new UnsupportedOperationException(
         s"Scoring with model of type ${model.getClass} in ${this.getClass} is not supported")
   }
+
+
+  /**
+   * Compute an optimized model (i.e. run the coordinate optimizer) for the current dataset.
+   *
+   * @return A (updated model, optimization state tracking information) tuple
+   */
+  override protected def trainModel(): (DatumScoringModel, OptimizationTracker) =
+    FixedEffectCoordinate.trainModel(dataset, optimizationProblem, featureShardId, None)
 }
 
 object FixedEffectCoordinate {
 
   /**
    * Train a new [[FixedEffectModel]] (i.e. run model optimization).
    *
-   * @param input The training dataset
+   * @param dataset The training dataset
    * @param optimizationProblem The optimization problem
    * @param featureShardId The ID of the feature shard for the training data
    * @param initialFixedEffectModelOpt An optional existing [[FixedEffectModel]] to use as a starting point for
    *                                   optimization
    * @return A new [[FixedEffectModel]]
    */
   private def trainModel[Function <: DistributedObjectiveFunction](
-      input: RDD[(UniqueSampleId, LabeledPoint)],
-      optimizationProblem: DistributedOptimizationProblem[Function],
-      featureShardId: FeatureShardId,
-      initialFixedEffectModelOpt: Option[FixedEffectModel]): FixedEffectModel = {
+    dataset: DataFrame,
+    optimizationProblem: DistributedOptimizationProblem[Function],
+    featureShardId: FeatureShardId,
+    initialFixedEffectModelOpt: Option[FixedEffectModel]): (FixedEffectModel, OptimizationTracker) = {
+
+    val rdd = dataset
+      .rdd
+      .map { row =>
+        val features = row.getAs[SparkVector](0)
+        val label = row.getDouble(1)
+
+        LabeledPoint(label, VectorUtils.mlToBreeze(features))
+      }
+    rdd.persist(StorageLevel.MEMORY_ONLY)
 
-    val newModel = initialFixedEffectModelOpt
+    val (glm, stateTracker) = initialFixedEffectModelOpt
       .map { initialFixedEffectModel =>
-        optimizationProblem.runWithSampling(input, initialFixedEffectModel.model)
+        optimizationProblem.runWithSampling(rdd, initialFixedEffectModel.model)
       }
-      .getOrElse(optimizationProblem.runWithSampling(input))
-    val updatedModelBroadcast = input.sparkContext.broadcast(newModel)
+      .getOrElse(optimizationProblem.runWithSampling(rdd))
 
-    new FixedEffectModel(updatedModelBroadcast, featureShardId)
+    rdd.unpersist()
+
+    (new FixedEffectModel(SparkSession.builder.getOrCreate.sparkContext.broadcast(glm), featureShardId),
```

Comment:
There should definitely be an `apply` method in `FixedEffectModel` to hide this method of generating models which takes a `RDD` as input.